azerothcore/azerothcore-wotlk
# feat(Core/SmartAI): Add more smart events

## Changes Proposed:
- Fix logic for NEAR_PLAYERS and its NEGATION counterpart
- Add SMART_EVENT_NEAR_UNIT, which checks for creature or gob entries in the vicinity, and does the action if the specified amount is met. This is used mainly by adds like Lava Spawn in Molten Core. If there's 16+ of them in the vicinity they all die.
- Add SMART_EVENT_AREA_CASTING, which sweeps the threat list for targets that are in range and casting. This allows us to interrupt more than just the current victim

## Issues Addressed:

- Progress towards https://github.com/chromiecraft/chromiecraft/issues/5705

## SOURCE:


## Tests Performed:

- Did testy
- 


## How to Test the Changes:


```
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 448);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(448, 0, 0, 0, 103, 0, 100, 0, 0, 448, 3, 40, 1200, 11, 60081, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On 3 Hoggers Nearby - Cast Explosion'),
(448, 0, 1, 0, 103, 0, 100, 0, 1, 185224, 3, 40, 1200, 11, 60081, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On 3 Object 185224 Nearby - Cast Explosion'),
(448, 0, 2, 0, 104, 0, 100, 0, 0, 20, 5000, 5000, 1200, 11, 11978, 2, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Nearby Casting - Cast Kick');
```
`.go c id 448`
`.npc add temp 448` 3 times to test if it's checking for Hoggers nearby
`.gob add temp 185224` 3 times to test if it's checking for this gob nearby
cast a spell nearby with another character also in the threat list, see action be done

## Known Issues and TODO List:




## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
